### DISCUSSION TITLE
Is Deno's implementation of dynamic import() ECMA-262 and test262 conformant?

### DISCUSSION THREAD

**Discussion by guest271314** [quickjs-ng/quickjs#629] [Q&A] [answered]
Consider this code from here [Deno dynamic import("./exports") throws module not found for "exports.js" dynamically created in the script](https://gist.github.com/guest271314/4637bb1288321256d7c14c72ebc81137). 

Note: *A* workaround for this is to use `import(import.meta.resolve("./exports.js"))`, though I don't think we should have to use specialized behaviour for *dynamic* `import()` for Deno.

Very simple question: Is Deno's implementation of dynamic `import()`, that throws when raw string specifier is used to dynamically `import()` a script created and written to the underlying file system *within the running script* ECMA-262 and test262 conformant?

```
// Deno dynamic import("./exports") throws module not found for "exports.js" dynamically created in the script
// dynamic_import_always_throws.js
// References: https://www.reddit.com/r/Deno/comments/18unb03/comment/kfsszsw/ https://github.com/denoland/deno/issues/20945
// Usage:
// deno run -A dynamic_import_always_throws.js
// bun run dynamic_import_always_throws.js
// node --experimental-default-type=module dynamic_import_always_throws.js
import { open, unlink } from "node:fs/promises";
const runtime = navigator.userAgent;
const encoder = new TextEncoder();
try {
  const script = `export default 1;`;
  // deno
  if (runtime.includes("Deno")) {
    await Deno.writeFile("exports.js", encoder.encode(script));
  }
  // node
  if (runtime.includes("Node")) {
    const dynamic = await open("./exports.js", "w");
    await dynamic.write(script);
    await dynamic.close();
  }
  // bun
  if (runtime.includes("Bun")) {
    await Bun.write("exports.js", encoder.encode(script));
  }
  const { default: module } = await import("./exports.js"); // Raw string specifier
  console.log({ module });
  console.log({ runtime });
} catch (e) {
  console.log("Deno always throws.");
  console.log({ runtime });
  console.trace();
  console.log(e.stack);
} finally {
  console.log("Finally");
  // node, bun
  if (runtime.includes("Node") || runtime.includes("Bun")) {
    await unlink("./exports.js");
  } // deno
  else if (runtime.includes("Deno")) {
    await Deno.remove("./exports.js");
  }
}
```

**Comment by bnoordhuis** [comment]
> Very simple question: Is Deno's implementation of dynamic import(), that throws when raw string specifier is used to dynamically import() a script created and written to the underlying file system within the running script ECMA-262 and test262 conformant?

Yes. ECMA-262 doesn't demand anything from implementers. I can't link directly to the right clauses for some reason<sup>1</sup> but it's `13.3.10 Import Calls` and then `16.2.1.8 HostLoadImportedModule`:

> The actual process performed is [host-defined](https://tc39.es/ecma262/#host-defined) [..]

It's arguably a quality-of-implementation issue but not a spec conformance issue.

<sup>1</sup> the closest I can link to is https://tc39.es/ecma262/#sec-ecmascript-language-scripts-and-modules

**Reply by guest271314** [reply]
So, if I understand your answer correctly, ECMA-262 officially allows implementers to throw *or* not throw for the same code?

**Reply by bnoordhuis** [reply]
Yes.

**Reply by guest271314** [reply]
Pardon, had a system crash. 

I am having a challenging time conceiving of a raw string specifier for dynamic `import()` test where Node.js, Bun, and every other JavaScript runtime that implements dynamic `import()` not throwing results in a PASS and Deno's implementation that does throw for raw string specifier where the script is created in the running script also results in PASS.

I can't think of any other interface or specification where that is the case.

**Reply by guest271314** [reply]
So, for example if that was a test in test262, how do you reconcile Deno's dynamic `import()` throwing for raw string specifier created in the running script throwing and every other JavaScript runtime not throwing for that case? How can a PASS be the result without special consideration for Deno?

**Reply by guest271314** [reply]
I guess this https://tc39.es/ecma262/#sec-HostLoadImportedModule does include the ambiguous term "or"

> An implementation of HostLoadImportedModule must conform to the following requirements:
> - The [host environment](https://tc39.es/ecma262/#host-environment) must perform [FinishLoadingImportedModule](https://tc39.es/ecma262/#sec-FinishLoadingImportedModule)(referrer, specifier, payload, result), where result is either a [normal completion containing](https://tc39.es/ecma262/#sec-completion-record-specification-type) the loaded [Module Record](https://tc39.es/ecma262/#sec-abstract-module-records) **or** a [throw completion](https://tc39.es/ecma262/#sec-completion-record-specification-type), either synchronously or asynchronously.
    
(Emphasis added)

**Reply by bnoordhuis** [reply, marked as answer]
That's right.

**Reply by guest271314** [reply]
Good to know. I can throw specifications out of the window as meaning anything.
